core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/QueryBindingSet.java

```diff
@@ -152,4 +190,220 @@ public boolean equals(Object other) {
 			return super.equals(other);
 		}
 	}
-}
+
+	private static class SmallMap
+			extends AbstractMap<String, Value> {
+		private final Object[] keysValues = new Object[SWITCH_TO_MAP];
+
+		public SmallMap() {
+		}
+
+		public SmallMap(SmallMap map) {
+			System.arraycopy(map.keysValues, 0, keysValues, 0, SWITCH_TO_MAP);
+		}
+
+		@Override
+		public int size() {
+
+			int size = 0;
+			for (int i = 0; i < keysValues.length; i = i + 2) {
+				if (keysValues[i] != null) {
+					size++;
+				}
+			}
+			return size;
+		}
+
+		private Iterator<Entry<String, Value>> iterator() {
+			return new Iterator<Entry<String, Value>>() {
```

Comment:
Do you think the iterator can be swapped out with something like Arrays.stream(keysValues).filter(something or other not null).map(some sort of cast or something).iterator()?

Reply:
No, because we need to step by two to get the value. This is because the keys are interspersed with the values. This could work if it was an array of bindings. Not when we have a mixed array of String and Values.

Reply:
Probably better to have two arrays or an array of bindings then. Will be much easier for the next developer to understand.